Cmake build for ACCESS3

anton-seaice opened the pull request:
This changes the CMake build in WW3 to support ACCESS-OM3. It includes the patches from https://github.com/COSIMA/access-om3/tree/main/WW3/patches.

There is a [library build ](https://github.com/ACCESS-NRI/ACCESS-OM3/pull/35) and a standalone build which adds the [ww3_shel executable](https://github.com/ACCESS-NRI/ACCESS-OM3/pull/66)

All the WW3 forks are a mess, so i just overwrote the CMakeLists.txt file. We could change it in the future to have a seperate CMakeLists.txt file for access if needed.

general FYI @chrisb13 @ezhilsabareesh8 @dougiesquire 

Closes #1

micaeljtoliveira commented on CMakeLists.txt:
```diff
@@ -1,68 +1,226 @@
+# to-do: make configurable 
+list(APPEND switches "CESMCOUPLED" "DIST" "MPI" "PR1" "FLX4" "ST6" "STAB0" "LN1" "NL1" "BT1" "DB1" "MLIM" "TR0" "BS0" "RWND" "WNX1" "WNT0" "CRX1" "CRT0" "O0" "O1" "O2" "O3" "O4" "O5" "O6" "O7" "O14" "O15" "IS2" "REF0" "NOGRB" "IC3")
+
+option(ACCESS3_WW3     "Use ACCESS3 dependencies and install ACCESS3 libraries" OFF)
```
Same as for other projects: consider appending the project name to the option name.

micaeljtoliveira commented on CMakeLists.txt:
```diff
@@ -1,68 +1,226 @@
-  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build type" FORCE)
-  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+if(NOT TARGET NetCDF::NetCDF_Fortran) #Access3Share probably already has NetCDF with PUBLIC interface
+  find_package(NetCDF 4.7.3 MODULE REQUIRED Fortran)
```
Is there a reason to not allow earlier versions of NetCDF? If so, would be good to have it written here as a comment.

micaeljtoliveira commented on CMakeLists.txt:
```diff
@@ -1,68 +1,226 @@
+  )
+  install(TARGETS ww3lib
+    EXPORT Ww3libTargets
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT AccessWW3Cmeps_RunTime NAMELINK_COMPONENT AccessWW3Cmeps_Development
```
Consider dropping `Cemps` from component name.

dougiesquire reviewed (commented):
Thanks @anton-seaice! Nothing much to add on @micaeljtoliveira's comments.

I do think the commit history should refer back to [`COSIMA/access-om3/WW3`](https://github.com/COSIMA/access-om3/tree/main/WW3) somewhere to maintain provenance and credit back to @micaeljtoliveira's original work.

anton-seaice commented:
> Looks like you are completely replacing the existing WW3 CMake build system. Are you planning to contribute this back upstream?

No plans at this point - the upstream CMake was hard to understand and stylistically quite different from how we write them. I couldn't see how we would reconcile build flags or switches with the upstream version.

We could make a totally seperate CMakeLists file to make this clearer?

micaeljtoliveira commented:
> No plans at this point - the upstream CMake was hard to understand and stylistically quite different from how we write them. I couldn't see how we would reconcile build flags or switches with the upstream version.
> 
> We could make a totally seperate CMakeLists file to make this clearer?

No, it's okay to leave it like this. Just wanted to understand the reasoning.

anton-seaice commented:
I think I addressed all the comments, - this can be rereviewed.

new builds

https://github.com/ACCESS-NRI/ACCESS-OM3/pull/35#issuecomment-2756448111
https://github.com/ACCESS-NRI/ACCESS-OM3/pull/66#issuecomment-2756311942

dougiesquire reviewed (approved):
Good with me. I still think [this](https://github.com/ACCESS-NRI/WW3/pull/2#pullrequestreview-2715704120) is a good idea, but I'll leave it up to you.
